core/src/test/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSourceTest.java

```diff
@@ -389,36 +389,67 @@ void pathUsingDeclarativeShadowAfterBuiltinShadowAfterGroup() {
     }
 
     @Test
-    void invalidPathMissingProperty() {
-        assertThatCode(() -> RootVariableSource.from(
+    void pathUsingBuiltinShadowAfterBuiltinShadow() {
+        var rootVariableSource = RootVariableSource.from(
                 planningSolutionMetaModel,
                 TestdataInvalidDeclarativeValue.class,
                 "shadow",
-                "missing",
+                "previous.previous",
```

Comment:
This is strange. 

- If I want to reference 5 values back, do I have to do previous.previous.previous.previous.previous? Also, this is done at build-time - how can I reference an arbitrary previous reference?
- What happens if any of the values in this chain are null?

Honestly, it seems to me that this should not be allowed.

Reply:
- I think we limit it to 2 somewhere, although that might only be for indirect facts
- Null -> it will be called, but there will be no edge in the graph

Reply:
Let's check if there's a limit. If not, let's introduce one. 2 sounds reasonable.